In [57]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

### HW-day1 Solutions


# Exercises

* Answer the following questions using reproducible Python code.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego?

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. Then using a for loop calculate the average money spent per customer for each area code.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

### Read in the Data

In [58]:
file_name = 'data/lego-sales.csv'
DF = pd.read_csv(file_name)
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

### What is in the Data?

The data has 14 variables and 620 observations. Categorical information includes first and last name, theme, subtheme, and image url. Numerical data includes age, year, pieces, price, and quantity. Some of the columsn are weird - like phone number it is a number, but really it's categorical. It seems like we will use it to figure out location. The set id and number even though they are numbers are not as useful

In [56]:
# Solutions
# 1 Describe what you see:
print(DF.columns)
print(DF.shape)
print(DF.dtypes)
DF.describe()

Index(['first_name', 'last_name', 'age', 'phone_number', 'set_id', 'number',
       'theme', 'subtheme', 'year', 'name', 'pieces', 'us_price', 'image_url',
       'quantity', 'money_spent', 'area_code'],
      dtype='object')
(620, 16)
first_name       object
last_name        object
age               int64
phone_number     object
set_id            int64
number           object
theme            object
subtheme         object
year              int64
name             object
pieces          float64
us_price        float64
image_url        object
quantity          int64
money_spent     float64
area_code        object
dtype: object


age        set_id    year       pieces    us_price    quantity  \
count  620.000000    620.000000   620.0   551.000000  620.000000  620.000000   
mean    34.356452  25124.982258  2018.0   254.206897   29.041613    1.437097   
std     11.276537    506.760720     0.0   357.738804   34.630623    0.712849   
min     16.000000  24548.000000  2018.0    13.000000    3.990000    1.000000   
25%     25.000000  24724.750000  2018.0    70.000000    9.990000    1.000000   
50%     33.000000  24804.500000  2018.0   114.000000   19.990000    1.000000   
75%     41.000000  25640.250000  2018.0   313.000000   29.990000    2.000000   
max     68.000000  26060.000000  2018.0  4634.000000  349.990000    5.000000   

       money_spent  
count   620.000000  
mean     41.084016  
std      54.042148  
min       3.990000  
25%       9.990000  
50%      19.990000  
75%      49.990000  
max     699.980000

### What are the most common first names?

Using the value_counts function on the first_name variable we can see that the three most common first names are Jackson, Joseph, and Jacob.

In [46]:
# 2 What are the three most common first names of purchasers?
DF['first_name'].value_counts()

first_name
Jackson      13
Jacob        11
Joseph       11
Michael      10
Lucas         8
             ..
Kelly         1
Katelynn      1
Stephanie     1
Miles         1
Talise        1
Name: count, Length: 211, dtype: int64

### What are the most common themes?

Again we can use value_counts on the theme variable and that the three most common themes are Star Wars, Nexo Knights, and tied for third is Gear and Mixels. These counts don't consider how many sets were bought. A more thorough analysis might consider when someone bought four of the same set.

In [47]:
# 3 What are the three most common themes of Lego sets purchased?
DF['theme'].value_counts()

theme
Star Wars                  75
Nexo Knights               64
Mixels                     55
Gear                       55
City                       45
Friends                    42
Ninjago                    38
Duplo                      35
Bionicle                   34
Creator                    25
Elves                      22
DC Comics Super Heroes     22
Marvel Super Heroes        19
Dimensions                 18
Disney Princess            15
The Angry Birds Movie      11
Architecture               10
Technic                    10
Minecraft                   9
Advanced Models             4
Ghostbusters                3
Seasonal                    3
Collectable Minifigures     3
Ideas                       2
Classic                     1
Name: count, dtype: int64

### In the Star Wars theme what is the most common subtheme?

Now we mask the theme variable so that we look only at "Star Wars" sets. Then we can do a value counts on the subtheme variable of the masked data. The most common subtheme in the Star Wars main theme is The Force Awakens.

In [48]:
#4 Among the most common theme of Lego sets purchased, what is the most common subtheme?

common_theme = "Star Wars"
mask = DF['theme']==common_theme
DF_common = DF[mask].copy()
DF_common['subtheme'].value_counts()

subtheme
The Force Awakens            15
Buildable Figures            11
Episode V                    10
MicroFighters                10
Battlefront                   7
Original Content              7
Episode III                   6
Seasonal                      3
Rebels                        3
Episode IV                    2
Ultimate Collector Series     1
Name: count, dtype: int64

### Below: A data frame for each of the age categories:

- "18 and under"
-  "19 - 25"
-  "26 - 35"
-  "36 - 50"
-  "51 and over"

### Then using the data frames we sum the quantity of sets purchased.

The age group from 36-50 purchased the most Lego sets with 313, followed by 26-35 year olds.

In [49]:
#5 Create data frames for each of the ages in the following categories: 
# "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

# create 5 masks
mask_u18 = (DF['age'] <= 18)
mask_19_25 = (DF['age'] >= 19) & (DF['age'] <=25)
mask_26_35 = (DF['age'] >= 26) & (DF['age'] <=35)
mask_36_50 = (DF['age'] >= 36) & (DF['age'] <=50)
mask_50u = (DF['age'] >= 50)

# create 5 data frames
DF_u18 = DF[mask_u18].copy()
DF_19_25 = DF[mask_19_25].copy()
DF_26_35 = DF[mask_26_35].copy()
DF_36_50 = DF[mask_36_50].copy()
DF_50u = DF[mask_50u].copy()


#6 Which age group has purchased the highest number of Lego sets.

print('under 18:')
print(DF_u18['quantity'].sum())
print('-----')

print('19-25')
print(DF_19_25['quantity'].sum())
print('-----')

print('26-35')
print(DF_26_35['quantity'].sum())
print('-----')

print('36-50')
print(DF_36_50['quantity'].sum())
print('-----')

print('over 50')
print(DF_50u['quantity'].sum())


under 18:
45
-----
19-25
174
-----
26-35
267
-----
36-50
313
-----
over 50
112


### Which age group spend the most money on Legos?

Again using our data frames above we print out the sum of the money spent. It looks like ages 36-50 also spent the most money.

In [50]:
#7 Which age group has spent the most money on Legos?

# Add a column for money spent.
DF['money_spent'] = DF['us_price']*DF['quantity']

# Then redo the masks
DF_u18 = DF[mask_u18].copy()
DF_19_25 = DF[mask_19_25].copy()
DF_26_35 = DF[mask_26_35].copy()
DF_36_50 = DF[mask_36_50].copy()
DF_50u = DF[mask_50u].copy()

# Then sum the money spent
print('under 18:')
print(DF_u18['money_spent'].sum())
print('-----')

print('19-25')
print(DF_19_25['money_spent'].sum())
print('-----')

print('26-35')
print(DF_26_35['money_spent'].sum())
print('-----')

print('36-50')
print(DF_36_50['money_spent'].sum())
print('-----')

print('over 50')
print(DF_50u['money_spent'].sum())

under 18:
948.5500000000001
-----
19-25
4939.26
-----
26-35
7576.33
-----
36-50
9532.870000000003
-----
over 50
3468.8799999999997


### Whic Lego theme has made the most money?

Here we select two columns: theme and money_spent. Then we take our data frame, for those columns, and group the data by theme. Finally we can sum the columns remaining (money_spent) and the sort the values so we can easily see which theme comes out on top. The Star Wars theme has made the most money for lego in this data.

In [51]:
#8 Which Lego theme has made the most money for Lego?

colummns = ['theme','money_spent']
DF[colummns].groupby('theme').sum().sort_values('money_spent',ascending=False)

money_spent
theme                               
Star Wars                    4447.83
Ninjago                      2279.41
City                         2211.35
Nexo Knights                 2209.07
Minecraft                    1549.90
Gear                         1533.21
Friends                      1279.44
Duplo                        1219.50
Elves                        1119.72
Ghostbusters                  879.95
Advanced Models               849.95
Marvel Super Heroes           839.70
Bionicle                      779.54
DC Comics Super Heroes        653.65
Architecture                  589.85
Technic                       572.88
Creator                       549.65
Dimensions                    484.76
The Angry Birds Movie         478.87
Mixels                        414.17
Disney Princess               237.80
Ideas                         209.97
Seasonal                       39.96
Classic                        29.99
Collectable Minifigures        11.97

### Which area-code spent the most money?

First we have to do some work to get the area code out of the phone number column. Then we can do a similar process to above, grouping by area code and then summing money spent. I looks like unknown regions purchased the most, but area code 956 had the highest amount spent for the ones where an area code was listed. There were actually a lot of nans in the phone number columns.

It might actually be more interesting to look at average money spent. Using a for loop we went through the data and calculate the average amount that each but dividing the money_spent by the number of observations for that column. In this case the average spent per customer is highest in area code 212. This is the Manhattan borough of New York City

In [52]:
#9 Which area code has spent the most money on Legos? 
# In the US the area code is the first 3 digits of a phone number.

DF['area_code'] = DF['phone_number'].apply(lambda x: str(x).split('-')[0])

columns = ['area_code','money_spent']
DF[columns].groupby('area_code').sum().sort_values('money_spent',ascending=False)

money_spent
area_code             
nan            3992.64
956             719.96
973             684.91
567             549.86
281             464.90
...                ...
216               9.99
509               9.99
308               9.99
325               9.98
636               4.99

[157 rows x 1 columns]

In [53]:
counts = DF['area_code'].value_counts()
counts

area_code
nan    92
281     9
712     9
567     9
441     9
       ..
509     1
419     1
787     1
520     1
402     1
Name: count, Length: 157, dtype: int64

In [54]:
spent = DF[columns].groupby('area_code').sum().sort_values('money_spent',ascending=False)

average_spent = []
for c in counts.items():
    average_spent.append(spent.loc[c[0]]['money_spent']/c[1])
    # if c[0] != 'nan':
    #     print(f'Area code {c[0]} spent an average of {spent.loc[c[0]]['money_spent']/c[1]}')

spent['average_spent'] = average_spent
spent.sort_values('average_spent',ascending=False)

money_spent  average_spent
area_code                            
212              24.99     359.980000
246              59.97     169.980000
856              79.96     144.990000
573              79.94     124.990000
813              85.97     114.985000
...                ...            ...
710             111.93       8.323333
949              84.97       7.490000
802              79.98       5.490000
617              59.97       4.990000
910              14.98       4.990000

[157 rows x 2 columns]